File: src/_1003.py
"""
LeetCode 1002 - Check If Word Is Valid After Substitutions
"""
from sys import *
from typing import *
from collections import *
from string import *


class Solution:
    def isValid(self, S: str) -> bool:
        s = []
        for ch in S:
            if ch == 'a':
                s += ch
            elif ch == 'b':
                if not s or s[-1] != 'a':
                    return False
                s += ch
            else:
                if not s or s[-1] != 'b':
                    return False
                s.pop(-1)
                s.pop(-1)
        return not s

File: src/_1003.java
import java.util.Stack;

/**
 * LeetCode 1003 - Check If Word Is Valid After Substitutions
 *
 * Greedy
 */
public class _1003 {

    public boolean isValid(String S) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < S.length(); i++) {
            char ch = S.charAt(i);
            if (ch == 'a') {
                stack.push(ch);
            } else if (ch == 'b') {
                if (stack.isEmpty() || stack.peek() != 'a') {
                    return false;
                }
                stack.push('b');
            } else {
                if (stack.isEmpty() || stack.peek() != 'b') {
                    return false;
                }
                stack.pop();
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
}

